File: Code Generation/results/python/eval_work/gemini_hard_count-beautiful-numbers.py
# === IMPORTS GERAIS (do dataset) ===
import random
import functools
import collections
import string
import math
import datetime

from typing import *
from functools import *
from collections import *
from itertools import *
from heapq import *
from bisect import *
from string import *
from operator import *
from math import *

inf = float('inf')

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def list_node(values: list):
    if not values:
        return None
    head = ListNode(values[0])
    p = head
    for val in values[1:]:
        node = ListNode(val)
        p.next = node
        p = node
    return head

def is_same_list(p1, p2):
    if p1 is None and p2 is None:
        return True
    if not p1 or not p2:
        return False
    return p1.val == p2.val and is_same_list(p1.next, p2.next)

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def tree_node(values: list):
    if not values:
        return None
    root = TreeNode(values[0])
    i = 1
    queue = deque()
    queue.append(root)
    while queue:
        node = queue.popleft()
        if i < len(values) and values[i] is not None:
            node.left = TreeNode(values[i])
            queue.append(node.left)
        i += 1
        if i < len(values) and values[i] is not None:
            node.right = TreeNode(values[i])
            queue.append(node.right)
        i += 1
    return root

def is_same_tree(p, q):
    if not p and not q:
        return True
    elif not p or not q:
        return False
    elif p.val != q.val:
        return False
    else:
        return is_same_tree(p.left, q.left) and is_same_tree(p.right, q.right)


# === MODEL RESPONSE ===
from functools import lru_cache

class Solution:
    def beautifulNumbers(self, l: int, r: int) -> int:
        # Maximum possible sum of digits for r < 10^9 is 9 * 9 = 81
        # The product of digits can be represented by the powers of 2, 3, 5, 7.
        # Since we only care if product % sum == 0, we can track the remainder
        # for a fixed sum S.
        
        def count_beautiful(n: int) -> int:
            if n == 0:
                return 0
            s_n = str(n)
            length = len(s_n)
            
            total_count = 0
            
            # We iterate over every possible sum of digits S from 1 to 81
            for target_sum in range(1, 82):
                @lru_cache(None)
                def dp(idx, current_sum, current_rem, is_less, is_started):
                    if current_sum > target_sum:
                        return 0
                    
                    if idx == length:
                        return 1 if (current_sum == target_sum and current_rem == 0) else 0
                    
                    res = 0
                    upper = int(s_n[idx]) if not is_less else 9
                    
                    for digit in range(upper + 1):
                        new_is_less = is_less or (digit < upper)
                        new_is_started = is_started or (digit > 0)
                        
                        if not new_is_started:
                            # Leading zeros
                            res += dp(idx + 1, 0, 0, new_is_less, False)
                        else:
                            # Calculate new remainder: (prod * digit) % target_sum
                            # To handle digit 0: if any digit is 0, prod is 0, so rem is 0.
                            new_sum = current_sum + digit
                            if digit == 0:
                                new_rem = 0
                            else:
                                new_rem = (current_rem * digit) % target_sum
                            
                            res += dp(idx + 1, new_sum, new_rem, new_is_less, True)
                    return res
                
                # We need a fresh start for each target_sum, though leading zero logic 
                # can be tricky. A simpler way:
                # If target_sum is fixed, the product must be divisible by target_sum.
                # Initial product remainder for a non-zero start is 1 % target_sum.
                
                @lru_cache(None)
                def solve(idx, current_sum, current_rem, is_less, is_started):
                    if current_sum > target_sum:
                        return 0
                    if idx == length:
                        return 1 if (current_sum == target_sum and current_rem == 0) else 0
                    
                    limit = int(s_n[idx]) if not is_less else 9
                    res = 0
                    for d in range(limit + 1):
                        nxt_less = is_less or (d < limit)
                        if not is_started and d == 0:
                            res += solve(idx + 1, 0, 0, nxt_less, False)
                        else:
                            nxt_sum = current_sum + d
                            if nxt_sum > target_sum: continue
                            
                            if not is_started: # First non-zero digit
                                nxt_rem = d % target_sum
                            else:
                                nxt_rem = (current_rem * d) % target_sum
                            
                            res += solve(idx + 1, nxt_sum, nxt_rem, nxt_less, True)
                    return res

                total_count += solve(0, 0, 0, False, False)
                solve.cache_clear()
                
            return total_count

        return count_beautiful(r) - count_beautiful(l - 1)

# === TEST FUNCTION (do dataset) ===
def check(candidate):
    assert candidate(l = 999,r = 999) == 1
    assert candidate(l = 1,r = 9) == 9
    assert candidate(l = 900,r = 1000) == 35
    assert candidate(l = 789,r = 890) == 35
    assert candidate(l = 500,r = 5000) == 1800
    assert candidate(l = 5,r = 50) == 13
    assert candidate(l = 100,r = 200) == 32
    assert candidate(l = 50,r = 100) == 9
    assert candidate(l = 10,r = 20) == 2
    assert candidate(l = 999,r = 1001) == 3
    assert candidate(l = 1,r = 999999999) == 670349658
    assert candidate(l = 990,r = 1000) == 3
    assert candidate(l = 100000000,r = 100000000) == 1
    assert candidate(l = 123,r = 456) == 110
    assert candidate(l = 1,r = 1) == 1
    assert candidate(l = 300,r = 500) == 70
    assert candidate(l = 99999,r = 100001) == 2
    assert candidate(l = 100,r = 150) == 19
    assert candidate(l = 1,r = 15) == 10
    assert candidate(l = 1000,r = 10000) == 3676
    assert candidate(l = 111,r = 222) == 33
    assert candidate(l = 1000,r = 1010) == 11
    assert candidate(l = 50,r = 55) == 1
    assert candidate(l = 100000,r = 200000) == 52892
    assert candidate(l = 100000,r = 150000) == 29004
    assert candidate(l = 999999,r = 1000001) == 2
    assert candidate(l = 123456,r = 654321) == 288118
    assert candidate(l = 1,r = 1000000) == 534358
    assert candidate(l = 333333,r = 444444) == 59841
    assert candidate(l = 90000000,r = 99999999) == 6259058
    assert candidate(l = 123456789,r = 2147483647) == 1428531442
    assert candidate(l = 7654321,r = 8765432) == 646518
    assert candidate(l = 1000000,r = 10000000) == 5311175
    assert candidate(l = 987654,r = 987664) == 3
    assert candidate(l = 10000000,r = 10001000) == 1001
    assert candidate(l = 10000000,r = 20000000) == 6285218
    assert candidate(l = 999999990,r = 999999999) == 2
    assert candidate(l = 999999,r = 999999) == 0
    assert candidate(l = 888888,r = 999999) == 57963
    assert candidate(l = 99999999,r = 100000000) == 1
    assert candidate(l = 5000,r = 75000) == 33900
    assert candidate(l = 1,r = 1000000000) == 670349659
    assert candidate(l = 777777,r = 888888) == 58602
    assert candidate(l = 987654321,r = 1000000000) == 7554239
    assert candidate(l = 999999,r = 9999999) == 5311174
    assert candidate(l = 500000,r = 510000) == 10001
    assert candidate(l = 987654321,r = 987654321) == 1
    assert candidate(l = 123456789,r = 234567890) == 73969442
    assert candidate(l = 2000000,r = 2010000) == 10001
    assert candidate(l = 100000,r = 1000000) == 486888
    assert candidate(l = 10000000,r = 10000000) == 1
    assert candidate(l = 555555555,r = 555555555) == 0
    assert candidate(l = 1000000,r = 1001000) == 1001
    assert candidate(l = 500000,r = 999999) == 270652
    assert candidate(l = 100000000,r = 300000000) == 134174863
    assert candidate(l = 1000000,r = 3000000) == 1175795
    assert candidate(l = 456789,r = 456789123) == 304015671
    assert candidate(l = 123456,r = 123567) == 46
    assert candidate(l = 56789,r = 567890) == 276722
    assert candidate(l = 88888888,r = 99999999) == 6902887
    assert candidate(l = 50000000,r = 60000000) == 6522427
    assert candidate(l = 500000000,r = 501000000) == 1000001
    assert candidate(l = 1234567,r = 12345678) == 6951395
    assert candidate(l = 12345,r = 67890) == 26696
    assert candidate(l = 98765432,r = 98765432) == 0
    assert candidate(l = 8000000,r = 8100000) == 100001
    assert candidate(l = 456789,r = 987654) == 286642
    assert candidate(l = 1234567,r = 8765432) == 4438756
    assert candidate(l = 555555,r = 666666) == 61137
    assert candidate(l = 5000000,r = 6000000) == 609303
    assert candidate(l = 50000,r = 55000) == 2738
    assert candidate(l = 1000000,r = 1010000) == 10001
    assert candidate(l = 800000,r = 900000) == 53367
    assert candidate(l = 500000000,r = 550000000) == 36195523
    assert candidate(l = 500000,r = 550000) == 30521
    assert candidate(l = 10000000,r = 100000000) == 57110376
    assert candidate(l = 111111,r = 222222) == 58091
    assert candidate(l = 100000000,r = 111111111) == 11111111
    assert candidate(l = 800000,r = 801000) == 1001
    assert candidate(l = 100000000,r = 200000000) == 66911398
    assert candidate(l = 500000,r = 600000) == 55997
    assert candidate(l = 999900,r = 1000100) == 124
    assert candidate(l = 123456789,r = 123456789) == 1
    assert candidate(l = 100000,r = 999999) == 486887
    assert candidate(l = 111111111,r = 222222222) == 73867777
    assert candidate(l = 8000000,r = 8999999) == 583871
    assert candidate(l = 987654,r = 987654) == 0
    assert candidate(l = 999000,r = 1001000) == 1335
    assert candidate(l = 123,r = 987654321) == 662795385
    assert candidate(l = 100000,r = 500000) == 216236
    assert candidate(l = 77777777,r = 88888888) == 6956761
    assert candidate(l = 5000000,r = 5010000) == 10001


# === INVOCATION ===
check(Solution().beautifulNumbers)
print("All tests passed")

File: Code Generation/results/cpp/eval_work/gemini_hard_count-beautiful-numbers.cpp
// task_id: count-beautiful-numbers
// difficulty: Hard

#include <cassert>
#include <iostream>
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <map>
#include <set>
#include <algorithm>
#include <cmath>
#include <climits>
#include <numeric>
#include <queue>
#include <stack>
#include <deque>
#include <functional>
#include <utility>
#include <tuple>

using namespace std;



// === MODEL RESPONSE ===
#include <iostream>
#include <vector>
#include <string>
#include <cstring>

using namespace std;

class Solution {
    // memo[index][tight_constraint][current_sum][current_product_rem][has_zero]
    // Max sum for 9 digits is 9 * 9 = 81.
    // However, product % sum depends on the target sum.
    // We iterate through all possible target sums (1 to 81).
    long long memo[12][2][82][82];
    string num_str;
    int target_sum;

    long long dp(int idx, bool tight, int sum, int prod_rem) {
        if (sum > target_sum) return 0;
        if (idx == num_str.size()) {
            return (sum == target_sum && prod_rem == 0) ? 1 : 0;
        }
        if (memo[idx][tight][sum][prod_rem] != -1) {
            return memo[idx][tight][sum][prod_rem];
        }

        long long count = 0;
        int limit = tight ? (num_str[idx] - '0') : 9;

        for (int d = 0; d <= limit; ++d) {
            int next_sum = sum + d;
            // If any digit is 0, the product becomes 0. 0 % target_sum is always 0.
            // We pass a special state for prod_rem if a 0 is encountered.
            // Since we only care if the final product is divisible by target_sum,
            // we can represent the "zero product" state as prod_rem = 0.
            int next_prod_rem;
            if (d == 0) {
                next_prod_rem = 0;
            } else {
                next_prod_rem = (prod_rem * d) % target_sum;
            }

            count += dp(idx + 1, tight && (d == limit), next_sum, next_prod_rem);
        }

        return memo[idx][tight][sum][prod_rem] = count;
    }

    long long solve_for_sum(long long n, int s) {
        num_str = to_string(n);
        target_sum = s;
        memset(memo, -1, sizeof(memo));
        // prod_rem starts at 1 because it's a multiplicative identity.
        // If a number starts with 0s (leading zeros), they shouldn't force 
        // the product to 0. However, the problem says "positive integer".
        // A simple way is to handle numbers digit by digit. 
        // To avoid leading zero issues with product, we use a separate DP or 
        // simply ensure we start multiplying only after the first non-zero digit.
        
        // Revised DP:
        return dp_with_zero_check(0, true, 0, 1, true);
    }

    long long memo2[12][2][82][82][2];
    long long dp_with_zero_check(int idx, bool tight, int sum, int prod_rem, bool leading_zeros) {
        if (sum > target_sum) return 0;
        if (idx == num_str.size()) {
            return (sum == target_sum && prod_rem == 0) ? 1 : 0;
        }
        if (memo2[idx][tight][sum][prod_rem][leading_zeros] != -1) {
            return memo2[idx][tight][sum][prod_rem][leading_zeros];
        }

        long long count = 0;
        int limit = tight ? (num_str[idx] - '0') : 9;

        for (int d = 0; d <= limit; ++d) {
            int next_sum = sum + d;
            int next_prod_rem;
            bool next_leading_zeros = leading_zeros && (d == 0);

            if (next_leading_zeros) {
                // Still in leading zeros, product doesn't change
                next_prod_rem = 1; 
            } else {
                if (d == 0) next_prod_rem = 0;
                else next_prod_rem = (prod_rem * d) % target_sum;
            }

            count += dp_with_zero_check(idx + 1, tight && (d == limit), next_sum, next_prod_rem, next_leading_zeros);
        }
        return memo2[idx][tight][sum][prod_rem][leading_zeros] = count;
    }

public:
    long long count_le(long long n) {
        if (n <= 0) return 0;
        long long total = 0;
        num_str = to_string(n);
        // Max possible sum for 10^9 - 1 is 9*9 = 81.
        for (int s = 1; s <= 81; ++s) {
            target_sum = s;
            memset(memo2, -1, sizeof(memo2));
            total += dp_with_zero_check(0, true, 0, 1, true);
        }
        return total;
    }

    long long beautifulNumbers(int l, int r) {
        return count_le(r) - count_le(l - 1);
    }
};

// === TEST HARNESS ===
int main() {
    Solution sol;

    assert((sol.beautifulNumbers(999, 999) == 1LL));
    assert((sol.beautifulNumbers(1, 9) == 9LL));
    assert((sol.beautifulNumbers(900, 1000) == 35LL));
    assert((sol.beautifulNumbers(789, 890) == 35LL));
    assert((sol.beautifulNumbers(500, 5000) == 1800LL));
    assert((sol.beautifulNumbers(5, 50) == 13LL));
    assert((sol.beautifulNumbers(100, 200) == 32LL));
    assert((sol.beautifulNumbers(50, 100) == 9LL));
    assert((sol.beautifulNumbers(10, 20) == 2LL));
    assert((sol.beautifulNumbers(999, 1001) == 3LL));
    assert((sol.beautifulNumbers(1, 999999999) == 670349658LL));
    assert((sol.beautifulNumbers(990, 1000) == 3LL));
    assert((sol.beautifulNumbers(100000000, 100000000) == 1LL));
    assert((sol.beautifulNumbers(123, 456) == 110LL));
    assert((sol.beautifulNumbers(1, 1) == 1LL));
    assert((sol.beautifulNumbers(300, 500) == 70LL));
    assert((sol.beautifulNumbers(99999, 100001) == 2LL));
    assert((sol.beautifulNumbers(100, 150) == 19LL));
    assert((sol.beautifulNumbers(1, 15) == 10LL));
    assert((sol.beautifulNumbers(1000, 10000) == 3676LL));
    assert((sol.beautifulNumbers(111, 222) == 33LL));
    assert((sol.beautifulNumbers(1000, 1010) == 11LL));
    assert((sol.beautifulNumbers(50, 55) == 1LL));
    assert((sol.beautifulNumbers(100000, 200000) == 52892LL));
    assert((sol.beautifulNumbers(100000, 150000) == 29004LL));
    assert((sol.beautifulNumbers(999999, 1000001) == 2LL));
    assert((sol.beautifulNumbers(123456, 654321) == 288118LL));
    assert((sol.beautifulNumbers(1, 1000000) == 534358LL));
    assert((sol.beautifulNumbers(333333, 444444) == 59841LL));
    assert((sol.beautifulNumbers(90000000, 99999999) == 6259058LL));
    assert((sol.beautifulNumbers(123456789, 2147483647) == 1428531442LL));
    assert((sol.beautifulNumbers(7654321, 8765432) == 646518LL));
    assert((sol.beautifulNumbers(1000000, 10000000) == 5311175LL));
    assert((sol.beautifulNumbers(987654, 987664) == 3LL));
    assert((sol.beautifulNumbers(10000000, 10001000) == 1001LL));
    assert((sol.beautifulNumbers(10000000, 20000000) == 6285218LL));
    assert((sol.beautifulNumbers(999999990, 999999999) == 2LL));
    assert((sol.beautifulNumbers(999999, 999999) == 0LL));
    assert((sol.beautifulNumbers(888888, 999999) == 57963LL));
    assert((sol.beautifulNumbers(99999999, 100000000) == 1LL));
    assert((sol.beautifulNumbers(5000, 75000) == 33900LL));
    assert((sol.beautifulNumbers(1, 1000000000) == 670349659LL));
    assert((sol.beautifulNumbers(777777, 888888) == 58602LL));
    assert((sol.beautifulNumbers(987654321, 1000000000) == 7554239LL));
    assert((sol.beautifulNumbers(999999, 9999999) == 5311174LL));
    assert((sol.beautifulNumbers(500000, 510000) == 10001LL));
    assert((sol.beautifulNumbers(987654321, 987654321) == 1LL));
    assert((sol.beautifulNumbers(123456789, 234567890) == 73969442LL));
    assert((sol.beautifulNumbers(2000000, 2010000) == 10001LL));
    assert((sol.beautifulNumbers(100000, 1000000) == 486888LL));
    assert((sol.beautifulNumbers(10000000, 10000000) == 1LL));
    assert((sol.beautifulNumbers(555555555, 555555555) == 0LL));
    assert((sol.beautifulNumbers(1000000, 1001000) == 1001LL));
    assert((sol.beautifulNumbers(500000, 999999) == 270652LL));
    assert((sol.beautifulNumbers(100000000, 300000000) == 134174863LL));
    assert((sol.beautifulNumbers(1000000, 3000000) == 1175795LL));
    assert((sol.beautifulNumbers(456789, 456789123) == 304015671LL));
    assert((sol.beautifulNumbers(123456, 123567) == 46LL));
    assert((sol.beautifulNumbers(56789, 567890) == 276722LL));
    assert((sol.beautifulNumbers(88888888, 99999999) == 6902887LL));
    assert((sol.beautifulNumbers(50000000, 60000000) == 6522427LL));
    assert((sol.beautifulNumbers(500000000, 501000000) == 1000001LL));
    assert((sol.beautifulNumbers(1234567, 12345678) == 6951395LL));
    assert((sol.beautifulNumbers(12345, 67890) == 26696LL));
    assert((sol.beautifulNumbers(98765432, 98765432) == 0LL));
    assert((sol.beautifulNumbers(8000000, 8100000) == 100001LL));
    assert((sol.beautifulNumbers(456789, 987654) == 286642LL));
    assert((sol.beautifulNumbers(1234567, 8765432) == 4438756LL));
    assert((sol.beautifulNumbers(555555, 666666) == 61137LL));
    assert((sol.beautifulNumbers(5000000, 6000000) == 609303LL));
    assert((sol.beautifulNumbers(50000, 55000) == 2738LL));
    assert((sol.beautifulNumbers(1000000, 1010000) == 10001LL));
    assert((sol.beautifulNumbers(800000, 900000) == 53367LL));
    assert((sol.beautifulNumbers(500000000, 550000000) == 36195523LL));
    assert((sol.beautifulNumbers(500000, 550000) == 30521LL));
    assert((sol.beautifulNumbers(10000000, 100000000) == 57110376LL));
    assert((sol.beautifulNumbers(111111, 222222) == 58091LL));
    assert((sol.beautifulNumbers(100000000, 111111111) == 11111111LL));
    assert((sol.beautifulNumbers(800000, 801000) == 1001LL));
    assert((sol.beautifulNumbers(100000000, 200000000) == 66911398LL));
    assert((sol.beautifulNumbers(500000, 600000) == 55997LL));
    assert((sol.beautifulNumbers(999900, 1000100) == 124LL));
    assert((sol.beautifulNumbers(123456789, 123456789) == 1LL));
    assert((sol.beautifulNumbers(100000, 999999) == 486887LL));
    assert((sol.beautifulNumbers(111111111, 222222222) == 73867777LL));
    assert((sol.beautifulNumbers(8000000, 8999999) == 583871LL));
    assert((sol.beautifulNumbers(987654, 987654) == 0LL));
    assert((sol.beautifulNumbers(999000, 1001000) == 1335LL));
    assert((sol.beautifulNumbers(123, 987654321) == 662795385LL));
    assert((sol.beautifulNumbers(100000, 500000) == 216236LL));
    assert((sol.beautifulNumbers(77777777, 88888888) == 6956761LL));
    assert((sol.beautifulNumbers(5000000, 5010000) == 10001LL));

    cout << "All tests passed for count-beautiful-numbers" << endl;
    return 0;
}
